engine/api/workflowtemplate/dao.go

```diff
@@ -0,0 +1,131 @@
+		return nil, sdk.WrapError(err, "Cannot get workflow template audits")
+	}
+
+	return awts, nil
+}
+
+// InsertInstance for workflow template in database.
+func InsertInstance(db gorp.SqlExecutor, wti *sdk.WorkflowTemplateInstance) error {
+	return sdk.WrapError(gorpmapping.Insert(db, wti), "Unable to insert workflow template relation %d with workflow %d",
+		wti.WorkflowTemplateID, wti.WorkflowID)
+}
+
+// UpdateInstance for workflow template in database.
+func UpdateInstance(db gorp.SqlExecutor, wti *sdk.WorkflowTemplateInstance) error {
+	return sdk.WrapError(gorpmapping.Update(db, wti), "Unable to update workflow template instance %d", wti.ID)
+}
+
+// DeleteInstanceNotIDAndWorkflowID removes all instances of a template where not id and workflow id equal in database.
+func DeleteInstanceNotIDAndWorkflowID(db gorp.SqlExecutor, id, workflowID int64) error {
+	_, err := db.Exec("DELETE FROM workflow_template_instance WHERE id != $1 AND workflow_id = $2", id, workflowID)
+	return sdk.WrapError(err, "Unable to remove all instances for workflow %d", workflowID)
+}
+
+// DeleteInstancesForWorkflowTemplateID removes all template instances by template id in database.
+func DeleteInstancesForWorkflowTemplateID(db gorp.SqlExecutor, workflowTemplateID int64) error {
+	_, err := db.Exec("DELETE FROM workflow_template_instance WHERE workflow_template_id = $1", workflowTemplateID)
+	return sdk.WrapError(err, "Unable to remove all instances for workflow template %d", workflowTemplateID)
+}
+
+// GetInstances returns all workflow template instances for given criteria.
+func GetInstances(db gorp.SqlExecutor, c CriteriaInstance) ([]*sdk.WorkflowTemplateInstance, error) {
+	wtis := []*sdk.WorkflowTemplateInstance{}
+
+	if _, err := db.Select(&wtis, fmt.Sprintf("SELECT * FROM workflow_template_instance WHERE %s", c.where()), c.args()); err != nil {
+		return nil, sdk.WrapError(err, "Cannot get workflow template instances")
+	}
+
+	return wtis, nil
+}
+
+// GetInstance returns a workflow template instance for given criteria.
+func GetInstance(db gorp.SqlExecutor, c CriteriaInstance) (*sdk.WorkflowTemplateInstance, error) {
+	wti := sdk.WorkflowTemplateInstance{}
+
+	if err := db.SelectOne(&wti, fmt.Sprintf("SELECT * FROM workflow_template_instance WHERE %s", c.where()), c.args()); err != nil {
+		if err == sql.ErrNoRows {
+			return nil, nil
+		}
+		return nil, sdk.WrapError(err, "Cannot get workflow template instance")
+	}
+
+	return &wti, nil
+}
+
+// InsertInstanceAudit for workflow template instance in database.
+func InsertInstanceAudit(db gorp.SqlExecutor, awti *sdk.AuditWorkflowTemplateInstance) error {
+	return sdk.WrapError(gorpmapping.Insert(db, awti), "Unable to insert audit for workflow template instance %d", awti.WorkflowTemplateInstanceID)
+}
+
+// GetInstanceAudits returns all workflow template instance audits for given criteria.
+func GetInstanceAudits(db gorp.SqlExecutor, c CriteriaInstanceAudit) ([]*sdk.AuditWorkflowTemplateInstance, error) {
```

Comment:
[]*sdk.AuditWorkflowTemplateInstance is not good... to man indirection for the escape analysis and the garbage collector

Reply:
done